engine/main.ml

```diff
@@ -224,39 +230,337 @@ let rec parse_lambda lsp =
+        | Int i, "==" -> Comparison (Eq, s1, i)
+        | Int i, "!=" -> Comparison (Nq, s1, i)
+        | _ -> assert false
+      in op, rem
+    | addint::var::tl when is_int_addition addint ->
+      begin match int_of_string_opt (BatString.rchop addint) with
+        | Some i -> Addition (i, var), tl
+        | None -> assert false
+      end
+    | other :: _ -> print ("Failure on "^other); assert false
     | [] -> assert false
   in
   match lsp with
   | (("true"|"false") as b)::")"::tl -> Bool (bool_of_string b), tl
-  | "("::((">"|"<"|">="|"<="|"=="|"!=") as bop)::tl -> print ("("^bop);  (* Comparison of bop * sexpr * int *)
-    let s1, s2, rem = advance_two_sexpr tl in
-    let op = match s2, bop with
-      | Int i, ">" -> Comparison (Gt, s1, i)
-      | Int i, "<" -> Comparison (Lt, s1, i)
-      | Int i, ">=" -> Comparison (Ge, s1, i)
-      | Int i, "<=" -> Comparison (Lt, s1, i)
-      | Int i, "==" -> Comparison (Eq, s1, i)
-      | Int i, "!=" -> Comparison (Nq, s1, i)
-      | _ -> assert false
-    in op, consume_last_paren rem
   | "("::rest ->
     let expr, rest = parse_special_form rest in
     expr, consume_last_paren rest
-  | x::tl ->
+  | x::tl -> 
      begin match int_of_string_opt x with
-       | Some i -> Int i, tl
+       | Some i -> print ("Int: "^x); Int i, tl
        | None ->
-          if x <> "" && x.[0] = '"' then
-            (assert (x.[String.length x - 1] = '"'); String x, tl)
-          else Var x, tl
+         assert (x <> ")") ;
+         if x <> "" && x.[0] = '"' then
+           (print ("String: "^x); assert (x.[String.length x - 1] = '"'); String x, tl)
+         else (print ("Var: "^x); Var x, tl)
      end
   | _ -> assert false
 
-let () =
+let parse_file filename =
   ignore target_example;
-  let target_example = BatFile.with_file_in Sys.argv.(1) BatIO.read_all in
+  let target_example = BatFile.with_file_in filename BatIO.read_all in
   let tk = tokenize target_example in
   let sexpr, tl = parse_lambda tk in
-  ignore sexpr;
-  if tl <> [] then Printf.eprintf "unparsed: %S\n%!" (String.concat " " tl)
+  if tl <> [] then Printf.eprintf "unparsed: %S\n%!" (String.concat " " tl); sexpr
+
+let example_ast =
+  parse_file Sys.argv.(1)
+
+module SMap = Map.Make(String)
+
+type constraint_tree =
+  | Leaf of pi list * target_blackbox
+  | Node of constraint_tree list
+and
+  pi = { var: accessor; op: piop } (* record of variable and constraint on that variable *)
+and
+  environment =
+  | Accessor of accessor
+  | Addition of variable * int
+  | Function of variable * constraint_tree 
+  | Catch of exitpoint * variable list * constraint_tree
+and
+  accessor =
+  | AcRoot of variable
+  | AcField of accessor * int
+  | AcTag of accessor * int
+and
+  piop =
+  | Tag of int
+  | NotTag of int
+  | Int of int
+  | NotInt of int
+  | Ge of int
+  | Gt of int
+  | Le of int
+  | Lt of int
+  | Eq of int
+  | Nq of int
+  | Isout of int
+  | Isin of int
```

Comment:
(maybe comment that Lambda doesn't actually have `isin`, we added it for negation-completeness)